Add Ufunc alias look up for appropriate numpy ufunc dispatching

VibhuJawa opened the pull request:
This PR closes https://github.com/rapidsai/cudf/issues/6921 by dispatching to appropriate cudf alias for numpy functions from the UFUNC_ALIASES dictionary :  
```python
 _UFUNC_ALIASES = {
    "power": "pow",
    "equal": "eq",
    "not_equal": "ne",
    "less": "lt",
    "less_equal": "le",
    "greater": "gt",
    "greater_equal": "ge",
    "absolute": "abs",
}
```

VibhuJawa commented on python/cudf/cudf/tests/test_array_ufunc.py:
```diff
@@ -27,7 +37,34 @@ def test_ufunc_cudf_series(np_ar_tup, func):
+    "np_ar_tup", [(np.random.random(100), np.random.random(100))]
+)
+@pytest.mark.parametrize(
+    "func", [np.logaddexp, np.fmax, np.fmod],
```
We currently only support a function b/w cudf series and cupy array using NumPy dispatch if we dispatch to cudf. We probably should not support either but we fail loudly in that case so I think this is fine. 

Unsure if we want to keep or change this behavior.  

That is currently below fails:
```python
>>> import cudf
i>>> import numpy as np
>>> import cupy as cp
>>> s = cudf.Series([1,2,3])
>>> ar = cp.array([1,2,3])
>>> np.greater(s,ar)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/nvme/0/vjawa/conda/envs/cudf_dev/lib/python3.7/site-packages/cudf/core/series.py", line 790, in __array_ufunc__
    cudf, cudf.Series, cupy, ufunc, inputs, kwargs
  File "/nvme/0/vjawa/conda/envs/cudf_dev/lib/python3.7/site-packages/cudf/utils/utils.py", line 488, in get_appropriate_dispatched_func
    return cudf_ser_func(*args, **kwargs)
  File "/nvme/0/vjawa/conda/envs/cudf_dev/lib/python3.7/site-packages/cudf/core/series.py", line 1670, in greater
    return self.gt(other)
  File "/nvme/0/vjawa/conda/envs/cudf_dev/lib/python3.7/site-packages/cudf/core/series.py", line 1664, in gt
    return self._binaryop(other, "gt", fill_value)
  File "/nvme/0/vjawa/conda/envs/cudf_dev/lib/python3.7/contextlib.py", line 74, in inner
    return func(*args, **kwds)
  File "/nvme/0/vjawa/conda/envs/cudf_dev/lib/python3.7/site-packages/cudf/core/series.py", line 1180, in _binaryop
    rhs = self._normalize_binop_value(rhs)
  File "/nvme/0/vjawa/conda/envs/cudf_dev/lib/python3.7/site-packages/cudf/core/series.py", line 1571, in _normalize_binop_value
    return self._column.normalize_binop_value(other)
  File "/nvme/0/vjawa/conda/envs/cudf_dev/lib/python3.7/site-packages/cudf/core/column/numerical.py", line 115, in normalize_binop_value
    other_dtype = np.min_scalar_type(other)
  File "<__array_function__ internals>", line 6, in min_scalar_type
TypeError: no implementation found for 'numpy.min_scalar_type' on types that implement __array_function__: [<class 'cupy.core.core.ndarray'>]

```
And below works. 
```python
>>> np.fmod(s,ar)
0    0
1    0
2    0
dtype: int64
```

kkraus14 replied:
I think we don't handle cupy arrays generally in binaryops and we likely should.

VibhuJawa replied:
Agreed, that we likely should, FWIW, pandas does support it with numpy. Happy to raise a feature request for it .  

```python
>>> s = pd.Series([1,2,3])
>>> ar = np.array([1,2,3])
>>> s.ge(ar)
0    True
1    True
2    True
dtype: bool
```

VibhuJawa replied:
Raised  https://github.com/rapidsai/cudf/issues/6974 to track/fix this as I feel this discussion should be out of the scope of this PR.

kkraus14 commented on python/cudf/cudf/core/series.py:
```diff
@@ -1642,6 +1651,9 @@ def le(self, other, fill_value=None, axis=0):
     def __le__(self, other):
         return self._binaryop(other, "le")
 
+    def less_equal(self, other, fill_value=None, axis=0):
```
docstring

kkraus14 commented on python/cudf/cudf/core/series.py:
```diff
@@ -1660,6 +1672,9 @@ def gt(self, other, fill_value=None, axis=0):
     def __gt__(self, other):
         return self._binaryop(other, "gt")
 
+    def greater(self, other, fill_value=None, axis=0):
```
docstring

VibhuJawa replied:
Resolved by https://github.com/rapidsai/cudf/pull/6973/commits/e968292218a9a30fb4c935feb5c15685bed00120

kkraus14 commented on python/cudf/cudf/core/series.py:
```diff
@@ -1678,6 +1693,9 @@ def ge(self, other, fill_value=None, axis=0):
     def __ge__(self, other):
         return self._binaryop(other, "ge")
 
+    def greater_equal(self, other, fill_value=None, axis=0):
```
docstring

VibhuJawa replied:
Resolved by https://github.com/rapidsai/cudf/pull/6973/commits/e968292218a9a30fb4c935feb5c15685bed00120

kkraus14 commented on python/cudf/cudf/core/series.py:
```diff
@@ -3840,6 +3858,9 @@ def abs(self):
     def __abs__(self):
         return self.abs()
 
+    def absolute(self):
```
docstring

VibhuJawa replied:
Resolved by https://github.com/rapidsai/cudf/pull/6973/commits/e968292218a9a30fb4c935feb5c15685bed00120

VibhuJawa commented on python/cudf/cudf/core/series.py:
```diff
@@ -1588,6 +1588,9 @@ def eq(self, other, fill_value=None, axis=0):
     def __eq__(self, other):
         return self._binaryop(other, "eq")
 
+    def equal(self, other, fill_value=None, axis=0):
```
So i read through pandas implementation, they seem to use the below mapping to do it. I think their solution is cleaner so I will switch to something similar.   

https://github.com/pandas-dev/pandas/blob/97153bf7544ddb8a8ba90376f1cc973dff8cae54/pandas/_libs/ops_dispatch.pyx#L22-L40

VibhuJawa replied:
Ignore this implementation for now.

VibhuJawa replied:
Switched this to a dictionary lookup when being dispatched . with commit https://github.com/rapidsai/cudf/pull/6973/commits/e968292218a9a30fb4c935feb5c15685bed00120

VibhuJawa commented on python/cudf/cudf/core/series.py:
```diff
@@ -1624,6 +1630,9 @@ def lt(self, other, fill_value=None, axis=0):
     def __lt__(self, other):
         return self._binaryop(other, "lt")
 
+    def less(self, other, fill_value=None, axis=0):
```
Resolved by https://github.com/rapidsai/cudf/pull/6973/commits/e968292218a9a30fb4c935feb5c15685bed00120

VibhuJawa commented on python/cudf/cudf/core/series.py:
```diff
@@ -1606,6 +1609,9 @@ def ne(self, other, fill_value=None, axis=0):
     def __ne__(self, other):
         return self._binaryop(other, "ne")
 
+    def not_equal(self, other, fill_value=None, axis=0):
```
Resolved by https://github.com/rapidsai/cudf/pull/6973/commits/e968292218a9a30fb4c935feb5c15685bed00120

VibhuJawa commented:
This should be ready for another reivew.

VibhuJawa commented:
This is failing when the LHS for `ufuncs` if the LHS for operations is a scalar. I should have caught this earlier.  we will need a generic fix for this. Let me think more about it. Shifting to WIP again, sorry for the back and forth.

VibhuJawa commented:
@kkraus14 , 

There are two solutions in my head to get a generic fix in:

a.  one is adding the functions as we have below : 
https://github.com/rapidsai/cudf/blob/598a14d820d47b7c3bfcb2bb3341b97a85317646/python/cudf/cudf/core/ops.py

b.  The other is writing a generic dispatching behavior where we check if the function being dispatched is a  **ufunc**, we do the following:  
  1. Check that if either LHS or RHS is of type cudf.Series
  2. We dispatch to the right function based on if the input is LHS or RHS. 

If we go with the second we might be able to remove https://github.com/rapidsai/cudf/blob/598a14d820d47b7c3bfcb2bb3341b97a85317646/python/cudf/cudf/core/ops.py to a generic solution but I am not sure if adding that code complexity is worth it to prevent repetition.

Please let me know what you think.

VibhuJawa commented:
@kkraus14 , 

This should be ready for another review. 

#### Code change summary 
1. Now for the below functions it checks if LHS/RHS is a series and dispatches an operation accordingly.  See [function](https://github.com/rapidsai/cudf/pull/6973/files#diff-edfe531e7c3a3ae863cc0e347232d7b000e0bc461601cddb0448180cae7574e4R501-R509
). 


Functions being dispatched:
```python
    "power": "pow",
    "equal": "eq",
    "not_equal": "ne",
    "less": "lt",
    "less_equal": "le",
    "greater": "gt",
    "greater_equal": "ge",
    "absolute": "abs",
```

2. Add check to ensure we error if out kwarg is being used with numpy operation. 

#### Tests Added
1. Added tests to cover scalars and cudf series when being used with NumPy array ufunc  operations.  (Scalar+Series, Series+Series, Series+Scalar)  
2. Added tests to ensure error in case of out kwarg 
3. Added tests for `bitwise_and`, `bitwise_or` and `bitwise_xor`


#### Future Cleanup: 
If we agree with the approach we can probably remove the operations in [cudf/core/ops.py ](https://github.com/rapidsai/cudf/blob/598a14d820d47b7c3bfcb2bb3341b97a85317646/python/cudf/cudf/core/ops.py) and just relying on a similar behavior for dispatch of those operations too. Let me know what you think.

brandon-b-miller commented on python/cudf/cudf/tests/test_array_ufunc.py:
```diff
@@ -23,11 +34,77 @@ def test_ufunc_cudf_series(np_ar_tup, func):
+    s_1, s_2 = cudf.Series(x), cudf.Series(y)
+    expect = func(x, y)
+    got = func(s_1, s_2)
+    if np.isscalar(expect):
```
Does this case ever occur if `x` and `y` are both `ndarray`s?

VibhuJawa replied:
nope, we can remove this.

brandon-b-miller commented on python/cudf/cudf/tests/test_array_ufunc.py:
```diff
@@ -23,11 +34,77 @@ def test_ufunc_cudf_series(np_ar_tup, func):
+        np.greater_equal,
+    ],
+)
+def test_ufunc_cudf_series_cudf_dispatch(np_ar_tup, func):
```
Usually we want to be able to set the seed for reproducibility. `np_ar_tup` should probably be updated to come from a data generating pytest fixture where the seed can be set, see a good example [here in the string tests.](https://github.com/rapidsai/cudf/blob/branch-0.18/python/cudf/cudf/tests/test_string.py#L55-L59).

VibhuJawa replied:
Resolved by https://github.com/rapidsai/cudf/pull/6973/commits/4be146433912fed09e094c81ea4881e7b7921df6

brandon-b-miller commented on python/cudf/cudf/tests/test_array_ufunc.py:
```diff
@@ -23,11 +34,77 @@ def test_ufunc_cudf_series(np_ar_tup, func):
+    s_1, s_2 = cudf.Series(x), cudf.Series(y)
+    expect = func(x, y)
+    got = func(s_1, s_2)
+    got = got.fillna(np.nan)
```
I think that this succeeds for a few of the comparison functions due to a bug in cuDF. 

```
x = cudf.Series(np.array([1, 2, np.nan])
y = cudf.Series(np.array([1, 2, np.nan])
```

The `nan` gets converted to a true `<NA>`, giving me

```
>>> y
0     1.0
1     2.0
2    <NA>
dtype: float64
```

Then doing a comparison between x and y would get me, for `>` for instance, 

```
>>> x > y
0    False
1    False
2    False
dtype: bool
```
The third entry should be an `<NA>` here. Currently when `fillna` is called then, it's a no-op for the cuDF series. But the numpy side of things gives us `array([False, False, False])` because it's comparing `NaN` rather than `<NA>`, so the comparison passes.

I think what we're supposed to be getting is 

```
>>> x > y
0    False
1    False
2    <NA>
dtype: bool
```

And then we'd ultimately be comparing `array([False, False, False])` to a cuDF series containing `[False, False, NaN]`. 

This should either be xfailed and be allowed to XPASS or have some conditional handling. 

@kkraus14 @shwina

shwina replied:
I'm not sure I'm following. What's the numpy result (`expect`) here?

If we want to test nan comparison behavior rather than null comparison behaviour, we should construct the cudf test input with `nan_as_null=False`. If we want to test null comparison behaviour, I don't think we can test against NumPy - we probably have to write our own expectation. Does that sound about right @brandon-b-miller?

brandon-b-miller replied:
basically we just have the wrong result in cuDF currently for comparison ops involving `<NA>`. 

```
>>> gsr = cudf.Series([1,2,None])
>>> gsr > gsr
0    False
1    False
2    False
dtype: bool
```

VibhuJawa replied:
@brandon-b-miller ,

For numpy we get below behavior which I think is what we want for comparison ops involving <NA>, are we planning on deviating from that behavior cause that will be concerning.   
```python
>>> import numpy as np
>>> ar = np.array([1,2,np.nan])
>>> ar > ar
array([False, False, False])
```

brandon-b-miller replied:
We want `nan` to follow the numpy comparison semantics. I think that works fine. 
```
>>> gsr = cudf.Series([1,2,None], dtype='float32')
>>> gsr
0     1.0
1     2.0
2    <NA>
dtype: float32
>>> gsr[2] = np.nan
>>> gsr
0    1.0
1    2.0
2    NaN
dtype: float32
>>> gsr > gsr
0    False
1    False
2    False
dtype: bool
```

But we want `<NA>` > `<NA>` to equal `<NA>` and currently it's returning `False`. I believe if it worked as such than the last element of the results wouldn't compare correctly.

VibhuJawa replied:
I think I get the confusion here, In this test I wanted to check that if we have `None` in our series it gets operated upon successfully for these binary ops and while writing the test I relied on the fact that we by default switch our nans to nulls, which i should be more explicit about.  

That said I still don't get why `<NA>` > `<NA>` should be equal to `<NA>`.  This is where my confusion is stemming from.  I think it should be equal to False.  

  ```python
>>> gsr = cudf.Series([1,None,2], dtype='float32')
>>> gsr > gsr
0    False
1    False
2    False
dtype: bool
````

VibhuJawa replied:
Had a chat with @brandon-b-miller , will add the comparison tests will nulls to a different test and mark them as xfail.

VibhuJawa replied:
Marked comparison operations as xfail, resolved by https://github.com/rapidsai/cudf/pull/6973/commits/4be146433912fed09e094c81ea4881e7b7921df6

brandon-b-miller commented:
Implementation looks fine. Had a few questions around the tests.

shwina commented on python/cudf/cudf/tests/test_array_ufunc.py:
```diff
@@ -23,11 +34,77 @@ def test_ufunc_cudf_series(np_ar_tup, func):
+    if np.isscalar(expect):
+        assert_eq(expect, got)
+    else:
+        assert_eq(expect, got.fillna(np.nan).to_array())
```
Is the call to `fillna()` here redundant with the one on line 81 above?

VibhuJawa replied:
yup, will remove it.

VibhuJawa replied:
Resolved by https://github.com/rapidsai/cudf/pull/6973/commits/4be146433912fed09e094c81ea4881e7b7921df6.

VibhuJawa commented:
@brandon-b-miller, Should be ready for a review now,. All reviews should have been addressed now.

brandon-b-miller reviewed (approved):
LGTM!
